Lib/test/test_urllib.py

```diff
@@ -1488,26 +1488,73 @@ def open_spam(self, url):
 class RequestTests(unittest.TestCase):
     """Unit tests for urllib.request.Request."""
 
+    def setUp(self):
+        self.Request = urllib.request.Request
+
+        self.headers = {
+            'User-agent': ('Mozilla/5.0 (X11; Linux i686; rv:10.0) '
+                           'Gecko/20100101 Firefox/10.0')
+        }
+
     def test_default_values(self):
-        Request = urllib.request.Request
-        request = Request("http://www.python.org")
+        request = self.Request("http://www.python.org")
         self.assertEqual(request.get_method(), 'GET')
-        request = Request("http://www.python.org", {})
+        request = self.Request("http://www.python.org", {})
         self.assertEqual(request.get_method(), 'POST')
 
     def test_with_method_arg(self):
-        Request = urllib.request.Request
-        request = Request("http://www.python.org", method='HEAD')
+        request = self.Request("http://www.python.org", method='HEAD')
         self.assertEqual(request.method, 'HEAD')
         self.assertEqual(request.get_method(), 'HEAD')
-        request = Request("http://www.python.org", {}, method='HEAD')
+        request = self.Request("http://www.python.org", {}, method='HEAD')
         self.assertEqual(request.method, 'HEAD')
         self.assertEqual(request.get_method(), 'HEAD')
-        request = Request("http://www.python.org", method='GET')
+        request = self.Request("http://www.python.org", method='GET')
         self.assertEqual(request.get_method(), 'GET')
         request.method = 'HEAD'
         self.assertEqual(request.get_method(), 'HEAD')
 
+    def test_request_with_headers(self):
+        """Verify headers if passed as a constructor argument."""
+        request = self.Request("http://www.python.org", headers=self.headers)
+        self.assertDictEqual(request.headers, self.headers)
+
+    def test_request_add_header(self):
+        """Verify headers if added through `Request.add_header()` method."""
+        request = self.Request("http://www.python.org")
+        request.add_header('User-agent', self.headers['User-agent'])
+        self.assertDictEqual(request.headers, self.headers)
+
+    def test_request_has_header_returns(self):
+        """Verify presence of headers is marked."""
+        request = self.Request("http://www.python.org", headers=self.headers)
+        self.assertTrue(request.has_header('User-agent'))
+        self.assertFalse(request.has_header('Content-type'))
```

Comment:
Test_urllib2.RequestHdrsTests.test_request_headers_methods already tests “has_header” against a constructor-supplied header field, and also has a negative test.